File: KSM/RT-comops-hrm-core/src/main/java/yowyob/comops/api/hrm/domain/model/Training.java
package yowyob.comops.api.hrm.domain.model;

import yowyob.comops.api.common.domain.model.BaseEntity;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.Objects;
import java.util.UUID;

public final class Training extends BaseEntity {

    private final UUID organizationId;
    private final UUID agencyId;
    private final String intitule;
    private final String organisme;
    private final LocalDate dateDebut;
    private final LocalDate dateFin;
    private final BigDecimal cout;
    private final Integer nbPlaces;
    private final String lieu;
    private final TrainingStatus status;

    private Training(UUID id, UUID tenantId, Instant createdAt, Instant updatedAt,
                     UUID organizationId, UUID agencyId, String intitule, String organisme,
                     LocalDate dateDebut, LocalDate dateFin, BigDecimal cout, Integer nbPlaces,
                     String lieu, TrainingStatus status) {
        super(id, tenantId, createdAt, updatedAt);
        this.organizationId = Objects.requireNonNull(organizationId);
        this.intitule = Objects.requireNonNull(intitule);
        this.dateDebut = Objects.requireNonNull(dateDebut);
        this.status = Objects.requireNonNull(status);
        this.agencyId = agencyId;
        this.organisme = organisme;
        this.dateFin = dateFin;
        this.cout = cout;
        this.nbPlaces = nbPlaces;
        this.lieu = lieu;
    }

    public static Training create(UUID tenantId, UUID organizationId, UUID agencyId, String intitule,
                                   String organisme, LocalDate dateDebut, LocalDate dateFin,
                                   BigDecimal cout, Integer nbPlaces, String lieu) {
        Instant now = Instant.now();
        return new Training(UUID.randomUUID(), tenantId, now, now, organizationId, agencyId,
                intitule, organisme, dateDebut, dateFin, cout, nbPlaces, lieu, TrainingStatus.PLANNED);
    }

    public static Training rehydrate(UUID id, UUID tenantId, Instant createdAt, Instant updatedAt,
                                      UUID organizationId, UUID agencyId, String intitule, String organisme,
                                      LocalDate dateDebut, LocalDate dateFin, BigDecimal cout, Integer nbPlaces,
                                      String lieu, TrainingStatus status) {
        return new Training(id, tenantId, createdAt, updatedAt, organizationId, agencyId,
                intitule, organisme, dateDebut, dateFin, cout, nbPlaces, lieu, status);
    }

    public Training start() {
        if (this.status != TrainingStatus.PLANNED) throw new IllegalStateException("Cannot start training in status " + this.status);
        return new Training(id(), tenantId(), createdAt(), Instant.now(), organizationId, agencyId,
                intitule, organisme, dateDebut, dateFin, cout, nbPlaces, lieu, TrainingStatus.IN_PROGRESS);
    }

    public Training complete() {
        if (this.status != TrainingStatus.IN_PROGRESS) throw new IllegalStateException("Cannot complete training in status " + this.status);
        return new Training(id(), tenantId(), createdAt(), Instant.now(), organizationId, agencyId,
                intitule, organisme, dateDebut, dateFin, cout, nbPlaces, lieu, TrainingStatus.COMPLETED);
    }

    public Training cancel() {
        if (this.status == TrainingStatus.COMPLETED) throw new IllegalStateException("Cannot cancel completed training");
        return new Training(id(), tenantId(), createdAt(), Instant.now(), organizationId, agencyId,
                intitule, organisme, dateDebut, dateFin, cout, nbPlaces, lieu, TrainingStatus.CANCELLED);
    }

    public UUID organizationId() { return organizationId; }
    public UUID agencyId() { return agencyId; }
    public String intitule() { return intitule; }
    public String organisme() { return organisme; }
    public LocalDate dateDebut() { return dateDebut; }
    public LocalDate dateFin() { return dateFin; }
    public BigDecimal cout() { return cout; }
    public Integer nbPlaces() { return nbPlaces; }
    public String lieu() { return lieu; }
    public TrainingStatus status() { return status; }
}

File: frontend/src/lib/types/hrm/training.ts
export type TrainingStatus = "PLANNED" | "IN_PROGRESS" | "COMPLETED" | "CANCELLED";

export type Training = {
  id: string;
  organizationId: string;
  agencyId: string | null;
  intitule: string;
  organisme: string | null;
  dateDebut: string;
  dateFin: string;
  cout: string;
  nbPlaces: number;
  lieu: string | null;
  status: TrainingStatus;
};

export type TrainingEnrollmentStatus = "ENROLLED" | "COMPLETED" | "CANCELLED";

export type TrainingEnrollment = {
  id: string;
  trainingId: string;
  employeeId: string;
  status: TrainingEnrollmentStatus;
  noteEvaluation: string | null;
  attestationFileId: string | null;
};

export type PlanTrainingInput = {
  agencyId?: string | null;
  intitule: string;
  organisme?: string | null;
  dateDebut: string;
  dateFin: string;
  cout: string;
  nbPlaces: number;
  lieu?: string | null;
};

export type EnrollEmployeeInput = {
  employeeId: string;
};

export type CompleteEnrollmentInput = {
  note?: string | null;
  attestationId?: string | null;
};

export type TrainingBudget = {
  id: string;
  organizationId: string;
  agencyId: string | null;
  annee: number;
  montantAlloue: string;
  montantEngage: string;
  montantRealise: string;
};

export type CreateTrainingBudgetInput = {
  agencyId?: string | null;
  annee: number;
  montantAlloue: string;
};

export type MontantInput = {
  montant: string;
};
